### PR TITLE
OpInfo for `nn.functional.avg_pool2d`

### PR DESCRIPTION
**Pull request by krshrimali**
Please see https://github.com/facebookresearch/functorch/issues/78

cc: @mruberry @zou3519

### CODE REVIEW

**Comment by mruberry** [comment]
@zou3519 and @heitorschueroff would you review this?

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2292,6 +2292,17 @@ def sample_inputs_leaky_relu(op_info, device, dtype, requires_grad):
                requires_grad=requires_grad)) for _ in range(1, N)]
     return tensors
 
+def sample_inputs_avgpool2d(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    cases = (((1, 3, 9, 9), (3, 3), (1, 1), (1, ), True, False, 2),
+             ((1, 1, 4, 4), (2, 2), (1, 1), (0, ), False, True, -2),
+             ((1, 2, 6, 6), (4, 4), (2, 2), (2, ), True, True, 3))
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:2300]: Let's add some more cases.
- Could we have one that is just a single kernel size `((1, 3, 9, 9),)`?
- avgpool_2d advertises that stride, padding, and kernel_size can be a single number, so we should test something like `((1, 3, 9, 9), 3, 1, 1)`
- Could we have a case where the two strides are not the same and another where the two paddings are not the same?
- We aren't testing division_override=None

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -2292,6 +2292,17 @@ def sample_inputs_leaky_relu(op_info, device, dtype, requires_grad):
                requires_grad=requires_grad)) for _ in range(1, N)]
     return tensors
 
+def sample_inputs_avgpool2d(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    cases = (((1, 3, 9, 9), (3, 3), (1, 1), (1, ), True, False, 2),
```
Comment (heitorschueroff) [inline torch/testing/_internal/common_methods_invocations.py:2298]: Can you add a comment describing what each element represents please

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -2292,6 +2292,17 @@ def sample_inputs_leaky_relu(op_info, device, dtype, requires_grad):
...
+             ((1, 1, 4, 4), (2, 2), (1, 1), (0, ), False, True, -2),
+             ((1, 2, 6, 6), (4, 4), (2, 2), (2, ), True, True, 3))
+
+    return [SampleInput(make_arg(input_shape), args=(kernel_size, stride, padding, ceil_mode,
```
Comment (heitorschueroff) [inline torch/testing/_internal/common_methods_invocations.py:2302]: nit: can you do this with a regular for loop instead of a list comprehension for readability.

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py:2302]: I really like list comprehensions. I think whether or not to do a list comprehension or do a for-loop should not be something we suggest in reviews because it's a matter of personal taste and there is no official PyTorch Style Guide.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Done, thanks for the suggestion @heitorschueroff!

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Done, thanks for the suggestion, @heitorschueroff!

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Thanks for taking a look @zou3519! I've addressed the suggestions in the latest commit. :)

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: I'm sorry, I saw your comment just now @zou3519! Since I had to add a single case without any args except kernel size as well, so I kept a `generator()` approach for now. Hope that's fine.

Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py]: > I really like list comprehensions. I think whether or not to do a list comprehension or do a for-loop should not be something we suggest in reviews because it's a matter of personal taste and there is no official PyTorch Style Guide.

List comprehensions are great for short one-liners, but I also agree that it is a matter of preferred style and since we don't enforce a style guide it is ok to ignore this suggestion.

Comment (zou3519) [reply torch/testing/_internal/common_methods_invocations.py]: Either style works, so this is fine :). I wouldn't say that one is nicer than the other because it depends on who is reading the code

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py]: Thanks for the valuable inputs, @zou3519 and @heitorschueroff! I appreciate it. :)

**Review by heitorschueroff — APPROVED** [review]
@krshrimali This is looking great, thank you for the contribution! I think there is room to explore a little more with the test cases, for instance, it looks like all test cases have size 1 for batch dimension. However, as a first PR adding OpInfo this is already good. Let's wait for the test to run. Thanks!

**Comment by krshrimali** [comment]
> @krshrimali This is looking great, thank you for the contribution! I think there is room to explore a little more with the test cases, for instance, it looks like all test cases have size 1 for batch dimension. However, as a first PR adding OpInfo this is already good. Let's wait for the test to run. Thanks!

Thanks, @heitorschueroff! I had two more cases in mind which I added in the recent commit:

- Since `stride` has a default value of `kernel_size`, passing an empty tuple (for `stride`) has been added in the cases. `None` can not be passed as JIT test will fail (`  Expected a value of type 'List[int]' for argument 'stride' but instead found type 'NoneType'.`).
- Multiple batch size, as also suggested by you. I think it's good for completion. (used 2 as the batch size, anything greater would be just extra load on the testing time, I thought).

cc: @zou3519 as well.

Hope this looks complete now. Thank you!

**Comment by krshrimali** [comment]
Gentle ping: @zou3519 and @heitorschueroff. The failures seem irrelevant to this PR, please let me know if there is anything else required here.

Thanks for your reviews and help.

**Comment by heitorschueroff** [comment]
> Gentle ping: @zou3519 and @heitorschueroff. The failures seem irrelevant to this PR, please let me know if there is anything else required here.
> 
> Thanks for your reviews and help.

LGTM! We appreciate your contribution.

**Comment by ngimel** [comment]
Reverting, causes flakiness (adaptive_avg_pool backward is nondeterministic)
